lib/s3routes/routes.js

```diff
@@ -0,0 +1,136 @@
+const errors = require('../errors');
+
+const routeGET = require('./routes/routeGET');
+const routePUT = require('./routes/routePUT');
+const routeDELETE = require('./routes/routeDELETE');
+const routeHEAD = require('./routes/routeHEAD');
+const routePOST = require('./routes/routePOST');
+const routeOPTIONS = require('./routes/routeOPTIONS');
+const routesUtils = require('./routesUtils');
+const routeWebsite = require('./routes/routeWebsite');
+
+const routeMap = {
+    GET: routeGET,
+    PUT: routePUT,
+    POST: routePOST,
+    DELETE: routeDELETE,
+    HEAD: routeHEAD,
+    OPTIONS: routeOPTIONS,
+};
+
+function checkUnsupportedRoutes(reqMethod, reqQuery, unsupportedQueries, log) {
+    const method = routeMap[reqMethod];
+    if (!method) {
+        return { error: errors.MethodNotAllowed };
+    }
+
+    if (routesUtils.isUnsupportedQuery(reqQuery, unsupportedQueries)) {
+        log.debug('encountered unsupported query');
+        return { error: errors.NotImplemented };
+    }
+
+    return { method };
+}
+
+function checkBucketAndKey(bucketName, objectKey, method, reqQuery,
+    isValidBucketName, log) {
+    // if empty name and request not a list Buckets
+    if (!bucketName && !(method === 'GET' && !objectKey)) {
+        log.debug('empty bucket name', { method: 'routes' });
+        return (method !== 'OPTIONS') ?
+            errors.MethodNotAllowed : errors.AccessForbidden
+               .customizeDescription('CORSResponse: Bucket not found');
+    }
+    if (bucketName !== undefined &&
+        isValidBucketName(bucketName) === false) {
+        log.debug('invalid bucket name', { bucketName });
+        return errors.InvalidBucketName;
+    }
+    if ((reqQuery.partNumber || reqQuery.uploadId)
+        && objectKey === undefined) {
+        return errors.InvalidRequest
+            .customizeDescription('A key must be specified');
+    }
+    return undefined;
+}
+
+function routes(req, res, params, logger) {
+    const {
```

Comment:
do you think we should do an assertion that each of these are here so the first request fails if anything is missing rather than waiting for some individual request later to fail?  also, since this is a major exported API, we should document the signature really well and assertions would help do that in addition to the jsdoc.